components/validator/l2_output_submitter.go

```diff
@@ -3,67 +3,222 @@ package validator
 import (
 	"context"
 	"errors"
+	"fmt"
 	"math/big"
 	_ "net/http/pprof"
+	"sync"
+	"time"
 
 	"github.com/ethereum/go-ethereum/accounts/abi"
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/log"
 
 	"github.com/kroma-network/kroma/bindings/bindings"
 	"github.com/kroma-network/kroma/components/node/eth"
 	"github.com/kroma-network/kroma/components/node/rollup"
+	"github.com/kroma-network/kroma/components/validator/metrics"
 	"github.com/kroma-network/kroma/utils"
+	"github.com/kroma-network/kroma/utils/service/txmgr"
 )
 
-// L2OutputSubmitter is responsible for submitting outputs
+// L2OutputSubmitter is responsible for submitting outputs.
 type L2OutputSubmitter struct {
-	log log.Logger
-	cfg Config
+	ctx    context.Context
+	cancel context.CancelFunc
 
-	l2ooContract *bindings.L2OutputOracleCaller
-	l2ooABI      *abi.ABI
+	cfg  Config
+	log  log.Logger
+	metr metrics.Metricer
+
+	l2ooContract    *bindings.L2OutputOracleCaller
+	l2ooABI         *abi.ABI
+	valpoolContract *bindings.ValidatorPoolCaller
+
+	submissionInterval *big.Int
+	l2BlockTime        *big.Int
+
+	txCandidatesChan chan<- txmgr.TxCandidate
+	submitChan       chan struct{}
+
+	wg sync.WaitGroup
 }
 
-// NewL2OutputSubmitter creates a new L2OutputSubmitter
-func NewL2OutputSubmitter(cfg Config, l log.Logger) (*L2OutputSubmitter, error) {
+// NewL2OutputSubmitter creates a new L2OutputSubmitter.
+func NewL2OutputSubmitter(parentCtx context.Context, cfg Config, l log.Logger, m metrics.Metricer, txCandidatesChan chan<- txmgr.TxCandidate) (*L2OutputSubmitter, error) {
+	ctx, cancel := context.WithCancel(parentCtx)
+
+	submitChan := make(chan struct{}, 1)
+
 	l2ooContract, err := bindings.NewL2OutputOracleCaller(cfg.L2OutputOracleAddr, cfg.L1Client)
 	if err != nil {
+		cancel()
 		return nil, err
 	}
 
 	parsed, err := bindings.L2OutputOracleMetaData.GetAbi()
 	if err != nil {
+		cancel()
 		return nil, err
 	}
 
+	valpoolContract, err := bindings.NewValidatorPoolCaller(cfg.ValidatorPoolAddr, cfg.L1Client)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	cCtx, cCancel := context.WithTimeout(ctx, cfg.NetworkTimeout)
```

Comment:
With this changes, please delete `cCancel()` in L78
```suggestion
	cCtx, cCancel := context.WithTimeout(ctx, cfg.NetworkTimeout)
	defer cCancel()
```